SWENT-GRP5/SOLVIT
# Fix: Location Dropdown & Service Requests

### Description:
This pull request optimizes the LocationDropdown component by adding debouncing, reducing the load on the API server when querying locations. It also refines the user interface in multiple areas, enhancing the app’s functionality and responsiveness.

- **What issue does it solve?** This PR addresses performance issues in the LocationDropdown by implementing debouncing, preventing excessive API requests, and improves user experience with enhancements to the date picker for the Creation and Edition of service requests and navigation in various screens.
- **What functionality does it add?** It adds debouncing to the LocationDropdown component, updates the query format for NominatimLocationRepository, improves the DatePicker field for ServiceRequests, and restores the back navigation arrow in CreateRequestScreen. Additionally, it includes fallback defaults for converting incomplete Firebase documents to ServiceRequest objects.

### Related Issues:
- #284 

### Changes Made:
- Updated `NominatimLocationRepository` query for enhanced API request accuracy, adding a "User-Agent" to the query.
- Added debouncing to `LocationDropdown` to avoid server overload, and used it both for the Service Requests and the `ProviderRegistrationScreen`.
- Restored the back navigation arrow in `CreateRequestScreen`, it seemed more logical and fit better in the UI.
- Modified the document-to-ServiceRequest conversion method to handle incomplete data by setting default values.
- Adapted tests for `LocationDropdown` and `DatePicker` to reflect recent modifications.
- Bonus (I find it very cool): completely modified the date picker field for ServiceRequests creation and editing screens.

### How to Test:
1. Log in as a seeker and while creating or editing a ServiceRequest, verify the functionality of the LocationDropdown (verify that debouncing is active, fewer requests should be sent to the server) and that the DatePicker works correctly.
2. Check that the back navigation arrow works properly in `CreateRequestScreen`.
3. Create an account as a Provider and while entering your information, verify that the LocationDropdown functions correctly
4. Run the updated AndroidTests to validate the functionality of the overall code.

---

### Checklist:

---

### As requested, here are some screenshots of the modifications made to the UI:

<img width="342" alt="Capture d’écran 2024-11-09 à 09 18 06" src="https://github.com/user-attachments/assets/3aaf5a0b-ba64-40f9-8274-8ccdb0be3f28">
<img width="336" alt="Capture d’écran 2024-11-09 à 09 20 37" src="https://github.com/user-attachments/assets/8bd37ecf-af5c-494c-8fa3-8a129e2b2116">
<img width="336" alt="Capture d’écran 2024-11-09 à 09 21 04" src="https://github.com/user-attachments/assets/be52622e-6509-4924-9671-e659da185742">
<img width="336" alt="Capture d’écran 2024-11-09 à 09 21 14" src="https://github.com/user-attachments/assets/82f1aebb-e33a-4bd8-88c5-e5f123e20b41">
